### PR TITLE
Added an async redis client implementation

### PR DESCRIPTION
**Pull request by euantorano**
This PR adds an `AsyncRedis` type. In order for this to work, the API has changed slightly (hence the version bump to `0.2.0`) as async procs can't capture `varargs` or `openarray`s.

In order to run the test procedures, use `nim c -d:testing -d:testasync -r src/redis.nim`. In the future I plan to add proper tests that can be ran using `nimble test` by CI.

It also adds documentation to the `docs` folder, and fleshes out the readme slightly.

Note that the async version will currently fail with Nim 0.17.0. THis seems to be caused by what I believe is a bug in the Nim stdlib when using `asyncnet.recvLine` or `asyncnet.recvLineInto` with unbuffered sockets - I believe [these](https://github.com/nim-lang/Nim/blob/master/lib/pure/asyncnet.nim#L537) two [lines](https://github.com/nim-lang/Nim/blob/master/lib/pure/asyncnet.nim#L544) should instead read `c = await recvFut` as otherwise the receive future may not have completed before the value is read (resulting in the error `Error: unhandled exception: Future still in progress.`). If anybody can confirm they get this same error, I will open an issue against [`nim-lang/Nim`](https://github.com/nim-lang/Nim).

### CODE REVIEW

**Comment by euantorano** [comment]
The copyright header in the source code also probably needs updating, but I'm not sure what you want it updating to. It currently reads the following:

```
#
#
#            Nim's Runtime Library
#        (c) Copyright 2012 Dominik Picheta
#
#    See the file "copying.txt", included in this
#    distribution, for details about the copyright.
#
```

**Comment by dom96** [comment]
Nice! But please don't commit the doc html files :)

File: src/redis.nim [hunk 1]
```diff
@@ -0,0 +1,1344 @@
...
+    raise newException(RedisError, "Server closed connection prematurely")
+
+  if line[0] == '-':
+    raise newException(RedisError, strip(line))
+  if line[0] != '+':
+    raiseInvalidReply('+', line[0])
+
+  result = line.substr(1) # Strip '+'
+
+proc readStatus(r: Redis | AsyncRedis): Future[RedisStatus] {.multisync.} =
```
Comment (dom96) [inline src/redis.nim]: Beautiful use of multisync :D

Comment (euantorano) [inline src/redis.nim]: Multisync is a godsend :smile:

**Comment by euantorano** [comment]
I'll remove those docs (and the reference from the README) now.

Any thoughts on the possible issue with `asyncdispatch.recvLine` in the stdlib?

I wouldn't say no to commit access. I don't use pedis a great deal, but do occasionally.

**Comment by dom96** [comment]
Huh, weird. I guess unbuffered sockets weren't ever really tested...

What you've written implies that this is a 0.17.0 problem and that a previous version is fine, is that so?

If not then go ahead and create a PR to fix this in the Nim repo.

**Comment by euantorano** [comment]
I'm not sure if it has ever worked to be honest. I'v only tested against 0.17.0. I'll create a PR in a couple of minutes to patch the issue.

**Comment by dom96** [comment]
But hrm, why are you using unbuffered sockets?

**Comment by euantorano** [comment]
Not sure, the original code was using unbuffered sockets, so I assumed you had a reason at the time :wink:

**Comment by euantorano** [comment]
Added some tests. The buffered sockets version doesn't need a patch to the stdlib, so this should be good to go. If there's a Nim Circle CI account, would you mind adding this project so that continuous integration will run the tests?

**Comment by dom96** [comment]
Done :) https://circleci.com/gh/nim-lang/redis/1

**Pull request merged**